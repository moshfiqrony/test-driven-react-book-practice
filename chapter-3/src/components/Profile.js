import React, { useState } from 'react';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';

function Profile(props) {
    const [firstName, setFirstName] = useState('')
    const [lastName, setLastName] = useState('')
    const [title, setTitle] = useState('')
    const [department, setDepartment] = useState('')
    const [pronoun, setPronoun] = useState('')
    const [manager, setManager] = useState(' ')
    const [timeZone, setTimeZone] = useState(' ')


    const handleCancel = () => {

    }


    const handleDone = () => {

    }

    return (
        <div className='container'>
            <div className="col-8 mx-auto mt-2">
                <div className="row">
                    <div className="col-6">
                        <label htmlFor='firstName'>First Name </label>
                        <input
                            name='firstName'
                            className='form-control'
                            type='text'
                            value={firstName}
                            onChange={(e) => setFirstName(e.target.value)}
                            placeholder='Enter your first name here'
                            data-testid='firstName'
                        />
                    </div>
                    <div className="col-6">
                        <label htmlFor='firstName'>Last Name </label>
                        <input
                            name='lastName'
                            className='form-control'
                            type='text'
                            value={lastName}
                            onChange={(e) => setLastName(e.target.value)}
                            placeholder='Enter your last name here'
                            data-testid='lastName'
                        />
                    </div>
                </div>
                <div className="row">
                    <div className="col-6">

                        <label htmlFor='firstName'>Title </label>
                        <input
                            name='title'
                            className='form-control'
                            type='text'
                            value={title}
                            onChange={(e) => setTitle(e.target.value)}
                            placeholder='Enter your first name here'
                            data-testid='title'
                        />
                    </div>
                    <div className="col-6">
                        <label htmlFor='firstName'>Department </label>
                        <input
                            name='department'
                            className='form-control'
                            type='text'
                            value={department}
                            onChange={(e) => setDepartment(e.target.value)}
                            placeholder='Enter your department here'
                            data-testid='department'
                        />
                    </div>
                </div>
                <label htmlFor='firstName'>Pronoun </label>
                <select
                    name='pronoun'
                    className='form-control'
                    type='text'
                    value={pronoun}
                    onChange={(e) => setPronoun(e.target.value)}
                    placeholder='Enter your pronoun here'
                    data-testid='pronoun'
                >
                    <option value='MALE'>He</option>
                    <option value='FEMALE'>She</option>
                    <option value='RATHER_NOT_SAY'>Rather not say</option>
                    <option value='NEUTRAL'>Neutral</option>
                </select>
                <label htmlFor='manager' >Manager </label>
                <Select
                    id='manager'
                    className='form-control'
                    value={manager}
                    onChange={(e) => setManager(e.target.value)}
                    data-testid='manager'
                >
                    <MenuItem value=' '>
                        Select One
                    </MenuItem>
                    <MenuItem value='user1'>
                        <img data-testid='manager-image' height='25' className='mr-2' src='https://img.icons8.com/officel/2x/user.png' />
                        User 1
                    </MenuItem>
                    <MenuItem value='user2'>
                        <img data-testid='manager-image' height='25' className='mr-2' src='https://img.icons8.com/officel/2x/user.png' />
                        User 2
                    </MenuItem>
                    <MenuItem value='user3'>
                        <img data-testid='manager-image' height='25' className='mr-2' src='https://img.icons8.com/officel/2x/user.png' />
                        User 3
                    </MenuItem>
                    <MenuItem value='user4'>
                        <img data-testid='manager-image' height='25' className='mr-2' src='https://img.icons8.com/officel/2x/user.png' />
                        User 4
                    </MenuItem>
                </Select>
                <label htmlFor='firstName'>Time Zone </label>
                <Select
                    className='form-control'
                    value={timeZone}
                    onChange={(e) => setTimeZone(e.target.value)}
                    data-testid='timeZone'
                >
                    <MenuItem value=' '>
                        Select One
                    </MenuItem>
                    <MenuItem value='asia/dhaka'>
                        Asia/Dhaka
                    </MenuItem>
                </Select>
                <button className='btn btn-light mt-2' onClick={handleCancel}>Cancel</button>
                <button className='btn btn-primary mt-2' onClick={handleDone}>Done</button>
            </div>
        </div>
    );
}

export default Profile;