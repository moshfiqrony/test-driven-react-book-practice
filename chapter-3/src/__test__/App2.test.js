import { render, fireEvent, waitFor, screen, wait } from '@testing-library/react'
import React from 'react'
import Profile from '../components/Profile';
import 'babel-polyfill'

const flushPromises = () => {
    return new Promise(resolve => {
      setTimeout(resolve, 0);
    })
  }

describe('Profile form test', () => {
    test('form is rendered properly?', () => {
        const { container } = render(<Profile />)
        expect(container).toMatchSnapshot();
    })


    test('onChange event and (type, name) is working properly for firstName?', () => {
        const { getByTestId } = render(<Profile />)
        const firstName = getByTestId('firstName');
        expect(firstName.name).toBe('firstName')
        expect(firstName.name).not.toBe('')
        expect(firstName.type).toBe('text')
        fireEvent.change(firstName, {target: {value: 'Moshfiq'}})
        expect(firstName.value).toBe('Moshfiq')
    })

    test('onChange event and (type, name) is working properly for lastName?', () => {
        const { getByTestId } = render(<Profile />)
        const lastName = getByTestId('lastName');
        expect(lastName.name).toBe('lastName')
        expect(lastName.name).not.toBe('')
        expect(lastName.type).toBe('text')
        fireEvent.change(lastName, {target: {value: 'Rony'}})
        expect(lastName.value).toBe('Rony')
    })

    test('onChange event and (type, name) is working properly for title?', () => {
        const { getByTestId } = render(<Profile />)
        const title = getByTestId('title');
        expect(title.name).toBe('title')
        expect(title.name).not.toBe('')
        expect(title.type).toBe('text')
        fireEvent.change(title, {target: {value: 'Software Engineer'}})
        expect(title.value).toBe('Software Engineer')
    })

    test('onChange event and (type, name) is working properly for department?', () => {
        const { getByTestId } = render(<Profile />)
        const department = getByTestId('department');
        expect(department.name).toBe('department')
        expect(department.name).not.toBe('')
        expect(department.type).toBe('text')
        fireEvent.change(department, {target: {value: 'Software Engineer'}})
        expect(department.value).toBe('Software Engineer')
    })

    test('onChange event and (type, name) is working properly for pronoun?', () => {
        const { getByTestId } = render(<Profile />)
        const pronoun = getByTestId('pronoun');
        expect(pronoun.name).toBe('pronoun')
        expect(pronoun.name).not.toBe('')
        expect(pronoun.type).toBe('select-one')
        fireEvent.change(pronoun, {target: {value: 'Software Engineer'}})
        expect(pronoun.value).toBe('MALE')
        fireEvent.change(pronoun, {target: {value: 'FEMALE'}})
        expect(pronoun.value).toBe('FEMALE')
    })

    test('onChange event and (type, name) is working properly for manager?', async() => {
        const {getByTestId} = render(<Profile />)
        fireEvent.mouseDown(screen.getAllByText('Select One')[0])
        fireEvent.click(screen.getByText('User 1'))
        expect(getByTestId('manager').querySelector('input').value).toBe('user1')
    })

    test('onChange event and (type, name) is working properly for timeZone?', () => {
        const {getByTestId} = render(<Profile />)
        fireEvent.mouseDown(screen.getAllByText('Select One')[1])
        fireEvent.click(screen.getByText('Asia/Dhaka'))
        expect(getByTestId('timeZone').querySelector('input').value).toBe('asia/dhaka')
    })
})
