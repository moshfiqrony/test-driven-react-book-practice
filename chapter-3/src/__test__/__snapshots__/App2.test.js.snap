// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile form test form is rendered properly? 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="col-8 mx-auto mt-2"
    >
      <div
        class="row"
      >
        <div
          class="col-6"
        >
          <label
            for="firstName"
          >
            First Name 
          </label>
          <input
            class="form-control"
            data-testid="firstName"
            name="firstName"
            placeholder="Enter your first name here"
            type="text"
            value=""
          />
        </div>
        <div
          class="col-6"
        >
          <label
            for="firstName"
          >
            Last Name 
          </label>
          <input
            class="form-control"
            data-testid="lastName"
            name="lastName"
            placeholder="Enter your last name here"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-6"
        >
          <label
            for="firstName"
          >
            Title 
          </label>
          <input
            class="form-control"
            data-testid="title"
            name="title"
            placeholder="Enter your first name here"
            type="text"
            value=""
          />
        </div>
        <div
          class="col-6"
        >
          <label
            for="firstName"
          >
            Department 
          </label>
          <input
            class="form-control"
            data-testid="department"
            name="department"
            placeholder="Enter your department here"
            type="text"
            value=""
          />
        </div>
      </div>
      <label
        for="firstName"
      >
        Pronoun 
      </label>
      <select
        class="form-control"
        data-testid="pronoun"
        name="pronoun"
        placeholder="Enter your pronoun here"
        type="text"
      >
        <option
          value="MALE"
        >
          He
        </option>
        <option
          value="FEMALE"
        >
          She
        </option>
        <option
          value="RATHER_NOT_SAY"
        >
          Rather not say
        </option>
        <option
          value="NEUTRAL"
        >
          Neutral
        </option>
      </select>
      <label
        for="manager"
      >
        Manager 
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline form-control"
        data-testid="manager"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="manager"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="manager"
          role="button"
          tabindex="0"
        >
          Select One
        </div>
        <input
          type="hidden"
          value=" "
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <label
        for="firstName"
      >
        Time Zone 
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline form-control"
        data-testid="timeZone"
      >
        <div
          aria-haspopup="listbox"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          role="button"
          tabindex="0"
        >
          Select One
        </div>
        <input
          type="hidden"
          value=" "
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <button
        class="btn btn-light mt-2"
      >
        Cancel
      </button>
      <button
        class="btn btn-primary mt-2"
      >
        Done
      </button>
    </div>
  </div>
</div>
`;
